### DISCUSSION TITLE
Extension of pipeflow for transient simulations and gas mixtures - internal structure changes

### DISCUSSION THREAD

**Discussion by dlohmeier** [e2nIEE/pandapipes#522] [Ideas]
As mentioned in some of the pandapipes community meetings, I would like to introduce a new internal structure in order to be able to better handle new simulation modes, such as transient heat transfer, dynamic controllers, gas mixtures, and maybe even optimization equations internally in pandapipes. 

The following aspects should at least be tackled:
- Create internal arrays (pit) that are necessary for the specific calculation mode, but don't add unnecessary columns there.
- Be able to transfer variables from one time step to the next internally without having to induce them into the tables from outside
- Better access to information on internal nodes of pipes that are split into several sections
- Ability to insert new equation types into the system matrix without too much effort or even adapting the whole build_system_matrix code

We might also think about:
- Adding solvers that can deal with other types of problems (non-linear, not Newton-Raphson, optimization ... )
- A better way to speed up the calculations, e.g. by making better use of global functions for all components, improved setting of variables that don't need to be updated from one time step to the other, ...

I see the following tasks here, please comment or update the task list if you have further ideas:
- [ ] Create a pandas table for internal node output
- [ ] Create component specific arrays for the internal structure that store values which are only processed within the component
- [ ] Make pit columns dynamic and create function to identify suiting pit columns for each simulation mode and setup
- [ ] Split up the build_system_matrix function to build the system matrix from different component models and for different modes

**Comment by AGeissler** [comment]
Almost a year later - has something happened, here? I am trying to find out, if/how pandapipes can be used at least «quasi-transient» (temperatures). In the code there are tests that have «transient=False» in the call e.g. to pipeflow. Sady, I could not find an example with «True». In the code-dokumentation of pipeflow there is a reference to **kwargs and that more information can be found in :func:`init_options` - however, I couldn't find this. Any help / suggestions etc. would be appreciated.

**Comment by dlohmeier** [comment]
Dear @AGeissler ,
unfortunately, the involvement in this topic has decreased drastically from different sides. If you take a look at the respective branch (transient_heat_transfer) inside the test/pipeflow_internals package, you can find some code in [test_transient.py](https://github.com/e2nIEE/pandapipes/blob/transient_heat_transfer/pandapipes/test/pipeflow_internals/test_transient.py). Currently, I cannot tell when this will be continued, but any contributions are highly welcome.

**Reply by AGeissler** [reply]
Thanks for the quick reply. Do you know if https://github.com/ERIGrid2/dhnsim_pandapipes works with the current pandapipes. That may be an option for me to look at, but as the last change in the repo was three years ago, I would not really be optimistic as pandapipes has seen quite a bit of development in the last years. If I / we can contribute remains to be seen, but of course if we can we will :) . Also, I see that there is a specific branch for _transient_ which I didn't have checked out (I was on develop, there is no test_transient.py there, that's why I missed it).

**Reply by dlohmeier** [reply]
Sorry, but I have no insights into this ERIGrid pandapipes coupling repo, I have no idea if the code is working.

**Reply by AGeissler** [reply]
Again, thanks. A quick look-see in the code seems to say that it would be necessary to have pandapipes w/ branch transient_heat_transfer checked out installed in the venv. Will current non-transient nets work with this? Or is this completely "development only"?
